cmd/api/src/database/migration/migrations/v5.2.1.sql

```diff
@@ -0,0 +1,21 @@
+-- Copyright 2023 Specter Ops, Inc.
+--
+-- Licensed under the Apache License, Version 2.0
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+-- SPDX-License-Identifier: Apache-2.0
+
+ALTER TABLE ONLY asset_group_selectors
+  DROP CONSTRAINT asset_group_selectors_name_key;
+
+ALTER TABLE ONLY asset_group_selectors
+  ADD CONSTRAINT asset_group_selectors_name_assetgroupid_key UNIQUE (name, asset_group_id);
```

Comment:
To be clear, both `ALTER TABLE` statements should be `ALTER TABLE IF EXISTS ONLY ...`, and the `DROP CONSTRAINT` should be `DROP CONSTRAINT IF EXISTS ...`. These are all painless changes that make the migration more idempotent.

Unfortunately, for adding the new constraint, we end up with something that's a bit more tricky. We can't simply run `ADD CONSTRAINT IF NOT EXISTS` and the alternative is to either try dropping it first (which could have performance implications in the future) or doing a pretty gross transaction that checks for existence and then runs the add if it doesn't exist. This one I'm less sure about how we should tackle it and whether it's worth doing so. Having all of our migrations being idempotent would be really good in theory, but it's also limited in practicality so how far we're willing to push the concept should probably be a point of further discussion.

Source for what it would take to make our constraint addition idempotent: https://stackoverflow.com/questions/6801919/postgres-add-constraint-if-it-doesnt-already-exist

Reply:
This migration would only run once. The assumption is that prior to this version `asset_group_selectors_name_assetgroupid_key`. That is to say this isn't a schema assertion that must be 100% idempotent. I think it's safe-enough to just have the constraint drop be idempotent.